rsfbclient-diesel/src/fb/transaction.rs

```diff
@@ -0,0 +1,94 @@
+//! Firebird transaction
+
+use super::connection::FbConnection;
+use diesel::connection::*;
+use diesel::result::Error::DatabaseError;
+use diesel::result::*;
+use diesel::QueryResult;
+use rsfbclient::SimpleTransaction as FbRawTransaction;
+use std::cell::{Cell, RefCell};
+
+/// Firebird transaction manager
+pub struct FbTransactionManager<'c> {
+    pub(crate) raw: RefCell<Option<FbRawTransaction<'c>>>,
+    depth: Cell<u32>,
+}
+
+impl<'c> FbTransactionManager<'c> {
+    pub fn new() -> Self {
+        FbTransactionManager {
+            raw: RefCell::new(None),
+            depth: Cell::new(0),
+        }
+    }
+}
+
+impl<'c> Default for FbTransactionManager<'c> {
+    fn default() -> Self {
+        FbTransactionManager::new()
+    }
+}
+
+impl<'c> TransactionManager<FbConnection<'c>> for FbTransactionManager<'c> {
+    fn begin_transaction(&self, conn: &FbConnection) -> QueryResult<()> {
+        let depth = self.depth.get() + 1;
+        if depth == 1 {
+            let conn = unsafe { conn.raw.as_ptr().as_mut().unwrap() };
```

Comment:
I've created a new tests called `transaction_depth_multithreaded` which causes random crashes, probably because of this unsafe block, but i still am not sure what exactly is happening.
I'll try to find a way to remove the unsafe here to see if the problem is fixed

Reply:
I'm trying resolve this bug too. My initial idea is use the [borrow_mut](https://doc.rust-lang.org/std/cell/struct.RefCell.html#method.borrow_mut), but this results in the [E0495 error](https://doc.rust-lang.org/error-index.html#E0495)

Reply:
Well, I believe the problem was not exactly in the unsafe block. I changed the `transaction_depth_multithreaded` test to use the `sql_query` instead the default query builder.

Unfortunately, now tarpaulin that is showing an error. I will check this.